File: src/HelloOgre3D/sandbox/state/MoveState.cpp
#include "MoveState.h"
#include "object/AgentObject.h"
#include "AgentTypeDef.h"
#include "ScriptLuaVM.h"

MoveState::MoveState(AgentObject* pAgent) : AgentState(pAgent)
{
	m_stateId = "MoveState";
}

MoveState::~MoveState()
{

}

void MoveState::OnEnter()
{
	
}

void MoveState::OnLeave()
{

}

std::string MoveState::OnUpdate(float dt)
{
	GetScriptLuaVM()->callFunction("Agent_MovingState", "u[AgentObject]i", m_pAgent, (int)dt);
	
	if (!m_pAgent->IsAnimReadyForMove())
	{
		return "";
	}

	auto pInput = m_pAgent->GetInput();
	if (pInput->isKeyDown(OIS::KC_2))
	{
		return "toShoot";
	}
	else if (pInput->isKeyDown(OIS::KC_3))
	{
		return ""; // ¼ÌÐøWalk
	}
	else if (pInput->isKeyDown(OIS::KC_4))
	{
		return "toDeath";
	}

	return "toIdle";
}

File: bin/res/scripts/states/MoveState.lua
-- MoveState --

--local m_pAgent = owner;
AgentState_OnEnter = function(agent)
    --print("Info==========>>>OnEnter")
end

AgentState_OnLeave = function(agent)
    --print("Info==========>>>OnLeave")
end

AgentState_OnUpdate = function(agent, deltaTime)
    local steeringForces = 0
    local deltaTimeInSec = deltaTime / 1000
    local stance = agent:getStanceType()
    if stance == SOLDIER_STAND then
        steeringForces = Soldier_CalculateSteering(agent, deltaTimeInSec)
    else
        steeringForces = Soldier_CalculateSlowSteering(agent, deltaTimeInSec)
    end
    agent:RequestState(SSTATE_RUN_FORWARD);
    Soldier_ApplySteering(agent, steeringForces, deltaTimeInSec)
    
    local pInput = agent:GetInput();
    if (pInput:isKeyDown(OIS.KC_2)) then
        return "toShoot";
    elseif pInput:isKeyDown(OIS.KC_3) then
        return ""; -- 继续Walk
    elseif (pInput:isKeyDown(OIS.KC_4)) then
        return "toDeath";
    end

    return "toIdle";
end

